ray-operator/controllers/common/ingress.go

```diff
@@ -0,0 +1,80 @@
+package common
+
+import (
+	"errors"
+	rayiov1alpha1 "github.com/ray-project/kuberay/ray-operator/api/raycluster/v1alpha1"
+	"github.com/ray-project/kuberay/ray-operator/controllers/utils"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const IngressClassAnnotationKey = "kubernetes.io/ingress.class"
+
+// BuildIngressForHeadService Builds the ingress for head service.
+// This is used to expose external an endpoint for external traffic.
+func BuildIngressForHeadService(cluster rayiov1alpha1.RayCluster) (*networkingv1.Ingress, error) {
+	labels := map[string]string{
+		RayClusterLabelKey: cluster.Name,
+		RayIDLabelKey:      utils.GenerateIdentifier(cluster.Name, rayiov1alpha1.HeadNode),
+	}
+
+	// Get ingress class name from rayCluster annotations. this is a required field to use ingress.
+	_, ok := cluster.Annotations[IngressClassAnnotationKey]
+	if !ok {
+		return nil, errors.New("can not find ingress class name from annotation")
+	}
+
+	// Copy other ingress configuration from cluster annotation
+	// This is to provide a generic way for user to customize their ingress settings.
+	annotation := map[string]string{}
+	for key, value := range cluster.Annotations {
+		annotation[key] = value
+	}
+
+	// The tricky thing is we need three ports. redis, dashboard, client. should we use /dashboard /redis to support that?
```

Comment:
I think the dashboard port is necessary, the redis/client-server port should be optional. We use ingress to expose port to internet access, but the redis/client-server is for internal network.

Reply:
Agree. Another tricky thing is redis/client are TCP/UDP that most ingress can not easily support well. I will modify the first version to support dashboard only.